ZCS-3326: Delegated admin cannot download the result of account migration CSV file

k-kato opened the pull request.

k-kato committed:
ZCS-3326: permitted a domain/delegated admin to access a bulkprovisioning result file

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Zimbra/zm-bulkprovision-store?pullRequest=1) <br/>All committers have signed the CLA.

fatkudu commented on src/java/com/zimbra/bp/BulkDownloadServlet.java:
```diff
@@ -31,7 +31,7 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOExc
             // check the auth token
             AuthToken authToken = ZimbraServlet.getAdminAuthTokenFromCookie(req, resp);
-            if (authToken == null || !authToken.isAdmin()) {
+            if (authToken == null || !AuthToken.isAnyAdmin(authToken)) {
```
I worry that this check may not be strict enough.  We need to be sure that admin rights are granted for the target resources and not some other resources.  For example, see how ZimbraSoapContext does more work to check what is allowed.

plobbes replied:
I also share @fatkudu 's concern.  There also seems to be a lack of sanitization on input parameters (e.g. fileId).

I'm wondering if this extension needs a full review to see how well Delegated Admin was considered in the initial design, however I don't have the time to dig into that right now.  You may want to discuss this more with @fatkudu and @grishick as I think they've both spent some time digging into similar issues in the past.

k-kato replied:
**Check for rights during importing account records**
If a delegated admin is allowed to access "Account Migration" page, the admin can run "Migration Wizard" without any right. (Note: createMigrationTask right is just for importing account DATA using IMAP)
On the other hand, some rigths like createAccount and createTopDomain are checked during "import account records" action on MBS.
If the admin does not have enough right(s), the action fails.

**Downloading CSV**
The result csv file is created when the import account record finishes, regardless of success or failure. 
The file name contains AccountId and random string defined as fileToken:

    [com.zimbra.bp.BulkImportAccounts]
    String fileToken = Double.toString(Math.random() * 100);
    String outSuccessFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
        File.separator, zsc.getAuthtokenAccountId(), fileToken);

An admin need to specify the random key as fileId like this:

    https://SERVER/service/extension/com_zimbra_bulkprovision/bulkdownload?action=getBulkFile&fileID=47.46394853406404&fileFormat=reportcsv

AccountId is also checked using AuthToken:

    [com.zimbra.bp.BulkDownloadServlet]
    } else if (pFileFormat.equalsIgnoreCase(ZimbraBulkProvisionExt.FILE_FORMAT_BULK_IMPORT_REPORT)) {
        bulkFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
                File.separator, authToken.getAccountId(), pFileId);
        clientFileName = "accounts_report.csv";

The file is deleted once the file is accessed:

    try {
        is.close();
        File file = new File(bulkFileName);
        file.delete();
    } catch (Exception e) {
        ZimbraLog.webclient.error(e);
    }

Or it is deleted by crontab by default.

    40 2 * * * /opt/zimbra/libexec/zmcleantmp

The random string works as one time token. It is difficult for a third party to guess the random token and the admin's AuthToken.
Of course, the file can be accessed by a third party if HTTPS connection is decrypted.

fatkudu replied:
@k-kato I think you are suggesting that the Admin Client will somehow do the security checks by stopping domain admins seeing the Account Migration page.  If true, that simply isn't good enough to stop bad actors who can use the REST API directly.  The server API handler needs to check that authenticated user has the rights to do what they are trying to do.

k-kato replied:
I don't think about UI side. I understand what you mean. Checking a right is important to handle API on a server.

BulkDownloadServlet.doGet handles two actions; (1)getSR (i.e. get search result) and (2)getBulkFile (i.e. get csv).
Regarding (1), rights are checked during the search action in SearchResults class. The pull request does not cause negative effect.

Regarding (2), I think there is no necessary right to download csv.
The reason is,
- any admin can call BulkImportAccountsRequest to import account records at this point,
- some rights like createAccount and createTop(Sub)Domain are checked during the import,
- an admin who calls BulkImportAccountsRequest should have a right to get the result csv file (Note: "right" is not a right defined on zimbra but a general word),
- only the admin who calls BulkImportAccountsRequest can access the result csv file as I mentioned in previous comment,
- if a bad actor uses the REST API without calling BulkImportAccountsRequest, a server just returns "File not found"

In addition, a right to run the import and/or to download a csv file is not defined on zimbra at this point.
If you think the extension must check rights to run the import and/or to download csv, we need to start from defining new right(s). It's better to address it in another ticket, I think.

fatkudu replied:
I would feel more comfortable if the same checks that BulkImportAccountsRequest does were done here too.  If I understand you correctly - this API gets the results of the BulkImportAccountsRequest - so I think it makes sense to require the same rights to get the result of the request as to make the request.

plobbes replied:
In addition to @fatkudu's comments, can we please filter/restrict data allowed in the request param ```fileId```? As it is right now, one could put in something like ```../../../blahblah``` or perhaps nulls, etc. and we wouldn't reject it. It seems like that should be to alphanumeric and it's length needs to be limited as well. We may want to see if there are others

k-kato replied:
@fatkudu I rethink what right is good for BulkImportAccountsRequest and for getting the result. New right(s) may need to be defined. I don't think it is suitable to check rights "createAccount", "createTopDomain" and "createSubDomain".

@plobbes I will add a logic to check the format and length of fileID.
Note:
- "action" and "fileFormat" parameter are used for switching an action. "fileID" is used for searching a target csv file. They are not used in a response from the server.
- If a wrong value is set in fileID, a server returns "File not found". Directory traversal is not caused by a wrong fileID in the current implementation.

      bulkFileName = String.format("%s%s_bulk_report_%s_%s.csv", LC.zimbra_tmp_directory.value(),
              File.separator, authToken.getAccountId(), pFileId);

      The file path with default localconfig is 
         /opt/zimbra/data/tmp/_bulk_report_[accountId]_[fileId].csv 
      Even if fileId is [xxx.csv/../../../../../yyy], the path is
         /opt/zimbra/data/tmp/_bulk_report_[accountId]_xxx.csv/../../../../yyy
      and then FileInputStream throws FileNotFoundException because "_bulk_report_[accountId]_xxx.csv" is not a directoy.

grishick reviewed (changes requested):
As far as I can see, this PR will let any delegated admin download a CSR and also download bulk provisioning files and bulk provisioning reports generated by other admins - all of these use cases are escalations of rights. If we need non-global admins to download bulk provisioning CSVs, I suggest the following course of action:
 1. move all CSR-related code out of this servlet into a separate servlet (we should do that anyway as a cleanup)
 2. decide whether a delegated admin should be able to download files generated by other admins (as long as both admins have sufficient permissions to list and get the generated accounts)
 3. depending on decision for #2 above: A map generated files to admin owners, so that admins cannot download each other's files. Or B - check that requesting admin has listAccount and getAccount permissions for all accounts inside a file being downloaded. The latter (B) maybe an overkill given that BP files may have a large number of accounts, so I suggest figuring out a way to assign files to admins and checking Admin.R_accessGAL right, which is, AFAIK, required to generate those bulk CSV files. Keep in mind that those files can have passwords in them.

k-kato committed:
ZCS-3326: Fixed as per review comments

- moved SR-related code to a new servlet
- modified authtoken check for downloading import results csv
- removed "action" param
- added format check for "fileId" param
- added return after sendError

k-kato commented:
1 => I modified as follows:
- moved CSR-related code into SearchResultsDownloadServlet.
- removed "action" param. (Note: zm-bulkprovision-admin-zimlet need to be modified)

I didn't write a code to check rights in SearchResultsDownloadServlet because necessary rights such as listAccount are checked during the search action.
- If an admin doesn't have right for search items, a generated csv doesn't contain them.
- A csv is not stored on a server. Other admins cannot download it.


2 & 3 => we need more discussion.
**Regarding a download of generated BP file from LDAP**
GenerateBulkProvisionFileFromLDAPRequest does not seem to be designed to support a delegated admin. It requires ldap URL, BindDN, BindPassword etc. like this:
```
    <GenerateBulkProvisionFileFromLDAPRequest xmlns="urn:zimbraAdminExt">
       <generatePassword>TRUE</generatePassword>
       <mustChangePassword>FALSE</mustChangePassword>
       <fileFormat>csv</fileFormat>
       <a n="zimbraGalLdapURL">ldap://SERVER:389</a>
       <a n="zimbraGalLdapFilter">(objectClass=zimbraAccount)</a>
       <a n="zimbraGalLdapSearchBase">dc=com</a>
       <a n="zimbraGalLdapBindDn">uid=zimbra,cn=admins,cn=zimbra</a>
       <a n="zimbraGalLdapBindPassword">XXXXXXXX</a>
    </GenerateBulkProvisionFileFromLDAPRequest>
```
It doesn't check accessGAL right during the process. Then we should deny access to a generated BP csv/xml by a delegated admin. Only a global admin should be able to download it.

**Regarding a download of import results csv**
BulkImportAccounts checks createAccount of a domain (and createTopDomain/createSubDomain as needed).
For example,
  - a delegated admin is admin@domain1.com
  - two domains domain1.com and domain2.com have already been created: 
  - rights of createAccount on domain1.com and domain2.com are checked
  - created (=imported) accounts: accoun1@domain1.com, account1@domain2.com

In this case, in BulkDownloadServlet, we don't know which domain(s) must be specified as the target of createAccount right. We cannot check the right without reading the csv file. 

If we choose option (A), we don't need to add more code to BulkDownloadServlet, I think.
  - A csv file name contains zimbraId of an admin who executed the import
  - BulkDownloadServlet specifies a csv file using zimbraId of an admin who tries to download it

Then a global/delegated admin cannot download an other admin's csv via the REST API.


**Others:**
- added format check for fileId, where @plobbes pointed out

grishick commented on src/java/com/zimbra/bp/BulkDownloadServlet.java:
```diff
@@ -21,89 +21,89 @@
-            } else {
-                return;
+            // check rights
+            if (isImportResultsDownload && !authToken.isAdmin()) {
```
I think this block can be removed. If the admin was able to generate the file, the same admin should be able to download it.

k-kato replied:
Removed the block.

grishick commented:
@k-kato I think it is OK to not let global admins download CSVs generated by other admins. If global admins really need to access these files, they can always use shell.

k-kato committed:
ZCS-3326: Fixed as per review comments

grishick reviewed (changes requested):
I'd like to request one more change. Please add a test to TestSearchResultsDownload.java to verify that if delegated admin with access to domainA and no access to domainB will not get results for accounts in domainB when searching across both domains and when searching specifically in domainB.

k-kato committed:
ZCS-3326: added tests of search results download by delegated admin

k-kato commented:
@grishick I added three tests.

k-kato commented:
This PR changes REST API. I submitted another PR on zm-bulkprovision-admin-zimlet.
https://github.com/Zimbra/zm-bulkprovision-admin-zimlet/pull/7

plobbes commented on src/java/com/zimbra/bp/SearchResultsDownloadServlet.java:
```diff
@@ -0,0 +1,76 @@
+            resp.setContentType("application/x-download");
+            resp.setHeader("Content-Disposition", "attachment; filename=search_result.csv");
+            SearchResults.writeSearchResultOutputStream(resp.getOutputStream(), query, domain, types, authToken);
+
```
nit: remove blank line

k-kato replied:
removed it

k-kato committed:
ZCS-3326: removed a blank line
